lib/dat-manager.js

```diff
@@ -0,0 +1,138 @@
+const Dat = require('dat-js')
+const root = require('./root-dir')
+const fs = require('fs')
+const EventEmitter = require('events')
+
+module.exports = class Manager extends EventEmitter {
+  constructor () {
+    super()
+    this.dats = []
+    this.loadAll((err, _dats) => {
+      if (err) return this.emit('error', err)
+      for (let dat of _dats) this.dats.push(this.watch(dat))
+      this.emit('update')
+    })
+  }
+  watch (dat) {
+    dat.on('update', () => this.emit('update'))
+    return dat
+  }
+  get () {
+    return this.dats
+  }
+  create (dir, cb) {
+    const dat = Dat({ dir })
+    dat.owner = true
+    dat.dir = dir
+    dat.share(err => {
+      if (err) return cb(err)
+      this.persist(dat, err => {
+        if (err) return cb(err)
+        this.dats.push(this.watch(this.activate(dat)))
+        this.emit('update')
+        cb()
+      })
+    })
+  }
+  download (key, cb) {
+    const dir = `${root}/${key}`
+    fs.mkdir(dir, err => {
+      if (err) return cb(err)
+      const dat = Dat({ dir, key })
+      dat.owner = false
+      dat.dir = dir
+      this.persist(dat, err => {
+        if (err) return cb(err)
+        dat.download()
+        this.dats.push(this.watch(this.activate(dat)))
+        this.emit('update')
+        cb()
+      })
+    })
+  }
+  load ({ key, owner, dir }) {
+    if (Buffer.isBuffer(key)) key = encoding.encode(key)
+    const dat = Dat({ key, dir })
+    dat.dir = dir
+    dat.owner = owner
+    if (dat.owner) {
+      dat.share(err => {
+        if (err) dat.emit('error', err)
+      })
+    } else {
+      dat.download(err => {
+        if (err) dat.emit('error', err)
+      })
+    }
+    return dat
+  }
+  loadAll (cb) {
+    const dats = []
+    db.createReadStream({
+      gt: ['archive', null],
+      lt: ['archive', undefined]
+    })
+    .on('data', data => {
+      dats.push(this.activate(this.load(data.value)))
+    })
+    .on('error', err => cb(err))
+    .on('end', () => cb(null, dats))
+  }
+  persist ({ key, owner, dir }, cb) {
+    db.put(['archive', key], {
+      key: encoding.encode(key),
+      owner,
+      dir
+    }, cb)
+  }
+  activate (dat) {
+    dat.open(err => {
+      if (err) throw err
+      dat.listStream = dat.archive.list({ live: true })
+      dat.listStream.on('data', entry => {
+        if (entry.name !== 'dat.json') return
+        collect(dat.archive.createFileReadStream('dat.json'), (err, raw) => {
+          if (err) return console.error('failed to read dat.json', err)
+          let json
+          try {
+            json = JSON.parse(raw.toString())
+          } catch (err) {
+            console.error('failed to parse dat.json', err)
+            return
+          }
+          dat.title = json.title
+          dat.emit('update')
+        })
+      })
+      dat.hyperStats = Stats({
+        archive: dat.archive,
+        db: sub(db, `${encoding.encode(dat.key)}-stats`)
+      })
+      dat.hyperStats.on('update', () => dat.emit('update'))
```

Comment:
not necessarily, hyperStats is somewhat independent of those events as it works up the archive's whole history instead of only watching for the "now". We will get duplicate update events still, but that as well should be solved once choo can do proper diffing as we should switch to immutable data structures underneath.